### PR TITLE
Allows search engines to be functions.

### PR DESCRIPTION
**Pull request by zappathustra**
Search engines can be functions, but `searchmenu` assumes they are only strings. This fixes it. Only engines whose functions do not raise an error are included in the menu.

### CODE REVIEW

File: searchmenu/init.lua [hunk 1]
```diff
@@ -62,8 +74,11 @@ new_mode("search-menu", {
         if search_engines then
             for name, uri in pairs(search_engines) do
                 local _title = lousy.util.escape(name)
-                local _uri = lousy.util.escape(string.format(uri, luakit.uri_encode(selection)))
-                table.insert(rows, 2, { "  " .. _title, " " .. _uri, uri = _uri })
+                local _uri = try_search_engine(uri, selection)
+                if _uri then
+                  _uri = lousy.util.escape(_uri)
```
Comment (aidanholm) [inline searchmenu/init.lua]: Indentation

File: searchmenu/init.lua [hunk 2]
```diff
@@ -14,6 +14,15 @@ local add_binds, add_cmds = modes.add_binds, modes.add_cmds
 local menu_binds = binds.menu_binds
 local new_mode = require("modes").new_mode
 
+local function try_search_engine (url, selection)
+    local ret
+    if type(url) == "string" then
+        ret = string.format(url, luakit.uri_encode(selection))
```
Comment (aidanholm) [inline searchmenu/init.lua]: Is there any reason not to just `return ...` inside the if statements? doesn't look like multiple return values would be an issue

File: searchmenu/init.lua [hunk 3]
```diff
@@ -14,6 +14,15 @@ local add_binds, add_cmds = modes.add_binds, modes.add_cmds
...
+    elseif pcall(url, selection) then
```
Comment (aidanholm) [inline searchmenu/init.lua]: only calling `url` once would be preferable

File: searchmenu/init.lua [hunk 4]
```diff
@@ -14,6 +14,15 @@ local add_binds, add_cmds = modes.add_binds, modes.add_cmds
 local menu_binds = binds.menu_binds
 local new_mode = require("modes").new_mode
 
+local function try_search_engine (url, selection)
```
Comment (aidanholm) [inline searchmenu/init.lua]: `url` isn't the best parameter name, maybe `search_engine` or `engine`

File: searchmenu/init.lua [hunk 5]
```diff
@@ -14,15 +14,15 @@ local add_binds, add_cmds = modes.add_binds, modes.add_cmds
...
-    if type(url) == "string" then
-        ret = string.format(url, luakit.uri_encode(selection))
-    elseif pcall(url, selection) then
-        ret = url(selection)
+local function try_search_engine (engine, selection)
+    if type(engine) == "string" then
+        return string.format(engine, luakit.uri_encode(selection))
+    else
+      local _, ret = pcall(engine, selection)
+      return ret
```
Comment (aidanholm) [inline searchmenu/init.lua]: This should be:
```lua
local ok, ret = pcall(...)
return ok and ret or nil
```

File: searchmenu/init.lua [hunk 6]
```diff
@@ -76,8 +76,8 @@ new_mode("search-menu", {
                 local _title = lousy.util.escape(name)
                 local _uri = try_search_engine(uri, selection)
                 if _uri then
-                  _uri = lousy.util.escape(_uri)
-                  table.insert(rows, 2, { "  " .. _title, " " .. _uri, uri = _uri })
```
Comment (aidanholm) [inline searchmenu/init.lua]: four spaces, not three :)

File: searchmenu/init.lua [hunk 7]
```diff
@@ -140,3 +140,5 @@ add_binds("normal", {
 add_cmds({
     { ":search-menu, :sm", "Select search engine from menu.", open_search_menu },
 })
+
+-- vim: et:sw=4:ts=8:sts=4:tw=80
```
Comment (aidanholm) [inline searchmenu/init.lua]: good catch :+1:

File: searchmenu/init.lua (hunk 5 above)
Comment (zappathustra) [reply searchmenu/init.lua]: If `pcall` fails, `ret` is `nil` anyway, so what's the difference?

Comment (aidanholm) [reply searchmenu/init.lua]: It'll be an error message, not nil

Comment (zappathustra) [reply searchmenu/init.lua]: Ah yes, stupid me !

**Comment by aidanholm** [comment]
Looks good! last request: please squash all the fixes into the first commit

**Comment by zappathustra** [comment]
Also, it would be nice to sort the search engines; for the moment, I find the unsorted list quite impractical. But I guess another PR is in order?

**Comment by aidanholm** [comment]
Thanks! And of course more PRs are always welcome :smile:

**Pull request merged**